### DISCUSSION TITLE
Table Widget: advance help

### DISCUSSION THREAD

**Discussion by joshikk** [ControlSystemStudio/phoebus#2030] [Q&A] [answered]
Hello,
I was thinking of using a table widget for compact display of acquired data.
The Display Buildet example is very limited, I could not find any documentation/elaborate examples with the table widget.
for ex: making VTable structure. having a sparsely populated table, for the NTTable type the table has to have all the elements.
Presently array PVs can be assigned to rows, can they be assigned to columns etc...
Can colouring of the cell be possible based upon the value etc...
Infact it should have all the table functionality that is available in core java
Is there any plan to add Tree Widget to phoebus.

Thank you in advance.

Regards,
Kuldeep

**Comment by kasemir** [comment]
See https://github.com/ControlSystemStudio/phoebus/wiki/Display-Builder-Table-Compatibility

**Comment by joshikk** [comment]
Hello Kay,
Thank you for the response.
The following two javadoc links in the online help are not working.
https://control-system-studio.readthedocs.io/en/latest/app/display/editor/doc/html/generated/org/csstudio/display/builder/model/Widget.html
https://control-system-studio.readthedocs.io/en/latest/app/display/editor/doc/html/generated/org/csstudio/display/builder/model/widgets/package-summary.html

The document says "a script can read a text file, read from a web service"
can you point to a sample program for the same.
I am not so conversant with Jython or embedded scripting of phoebus, so It
will be helpful if the online documentation mentions of the other
external/internal modules that can be called/used in the embedded scripts
for number processing, statistical analysis, os and other support.

Regards,
Kuldeep

**Comment by kasemir** [comment]
Links to the API only work in the complete help as it's bundled in the application, not with the standalone web help

![Screen Shot 2021-10-14 at 7 44 36 AM](https://user-images.githubusercontent.com/1932421/137311229-b1009531-50d6-497b-84a0-8446fd81b86f.png)

![Screen Shot 2021-10-14 at 7 44 59 AM](https://user-images.githubusercontent.com/1932421/137311250-d65a2a9c-b680-43ba-ac33-85e59e17736b.png)

In the examples, check scripts, web service, i.e., https://github.com/ControlSystemStudio/phoebus/tree/master/app/display/model/src/main/resources/examples/webservice, for an example that reads text from a web service and then displays that.

The thing about scripts in general: They are an advanced feature for those users who want to do dig into the API, i.e. read the source code, study all the examples, to accomplish some special behavior that is fundamentally not supported by the display, but can be done because it's an open system where you do have the freedom to dig into the source code and call anything in there from a script. But that also means you do have to look into the source code, and you might have to update the script as the displays builder source code changes, 
https://github.com/ControlSystemStudio/phoebus/wiki/Display-Builder-Script-Compatibility

**Comment by kasemir** [comment, marked as answer]
> I am not so conversant with Jython or embedded scripting of phoebus,

Then maybe simply don't use it. Most of the other display tools (EDM, MEDM, ... ) don't have a table widget because Channel Access doesn't provide table data. PVAccess now does, but only if you write your own custom server to generate it.

In Phoebus, developers on the bleeding edge can start to show table data if they put in their own custom script code to generate the table data, but as you found out by now, that's not easy. It basically means extending the display builder (via a script), it's way beyond simply using the features that it offers out of the box.
